packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol

```diff
@@ -116,4 +736,370 @@ library ConnextUtils {
+    if (keccak256(_args.executeArgs.params.callData) == EMPTY) {
+      // no call data, send funds to the user
+      _transferAssetFromContract(_adopted, _args.executeArgs.params.to, _amount, _args.wrapper);
+    } else {
+      // execute calldata w/funds
+      _transferAssetFromContract(_adopted, address(_args.executor), _amount, _args.wrapper);
+      _args.executor.execute(
+        _transferId,
+        _amount,
+        payable(_args.executeArgs.params.to),
+        _adopted,
+        _reconciled
+          ? LibCrossDomainProperty.formatDomainAndSenderBytes(
+            _args.executeArgs.params.originDomain,
+            _args.executeArgs.originSender
+          )
+          : LibCrossDomainProperty.EMPTY_BYTES,
+        _args.executeArgs.params.callData
+      );
+    }
+  }
+
+  /**
+   * @notice Execute liquidity process used when calling `execute`
+   * @dev Need this to prevent stack too deep
+   */
+  function _handleExecuteLiquidity(
+    bytes32 _transferId,
+    bool _isFast,
+    ExecuteLibArgs calldata _args,
+    mapping(bytes32 => address[]) storage _routedTransfers,
+    mapping(address => mapping(address => uint256)) storage _routerBalances,
+    mapping(bytes32 => IStableSwap) storage _adoptedToLocalPools,
+    mapping(bytes32 => address) storage _canonicalToAdopted,
+    RouterPermissionsManagerInfo storage _routerInfo
+  ) private returns (uint256, address) {
+    uint256 toSwap = _args.executeArgs.amount;
+    uint256 pathLen = _args.executeArgs.routers.length;
+    if (_isFast) {
+      // this is the fast liquidity path
+      // ensure the router is whitelisted
+
+      // calculate amount with fast liquidity fee
+      toSwap = _getFastTransferAmount(
+        _args.executeArgs.amount,
+        _args.LIQUIDITY_FEE_NUMERATOR,
+        _args.LIQUIDITY_FEE_DENOMINATOR
+      );
+
+      // TODO: validate routers signature on path / transferId
+
+      // store the routers address
+      _routedTransfers[_transferId] = _args.executeArgs.routers;
+
+      // for each router, assert they are approved, and deduct liquidity
+      uint256 routerAmount = toSwap / pathLen;
+      for (uint256 i; i < pathLen; ) {
+        // while theres no way for a router to have sufficient liquidity
+        // if they have never been approved, this check ensures they weren't
+        // removed from the whitelist
+        if (!_routerInfo.approvedRouters[_args.executeArgs.routers[i]]) {
+          revert ConnextUtils__execute_notSupportedRouter();
+        }
+
+        // decrement routers liquidity
+        _routerBalances[_args.executeArgs.routers[i]][_args.executeArgs.local] -= routerAmount;
+
+        unchecked {
+          i++;
+        }
+      }
+    } else {
+      // this is the slow liquidity path
+
+      // save a dummy address for the router to ensure the transfer is not able
+      // to be executed twice
+      // TODO: better ways to enforce this safety check?
+      _routedTransfers[_transferId] = [address(1)];
```

Comment:
did you have any thoughts on this TODO as you were refactoring the code?

Reply:
I think we can remove this and use `if (_transferRelayer[transferId] != address(0))` in the `_executeSanityChecks` to check that the transfer has not already been executed.